### ISSUE TITLE
[Feature Request] Support `MeterRegistry` Bean for Client in Spring integration

### ISSUE THREAD

**Issue by kev1n11** [line/armeria#3101]
Currently, the static global `MeterRegistry` is used for client metrics as per https://github.com/line/armeria/blob/3d30b065ee1a1542f54ffe3a03dea6b4cc36c959/core/src/main/java/com/linecorp/armeria/client/ClientFactoryOptions.java#L179

```
/**
     * The {@link MeterRegistry} which collects various stats.
     */
    public static final ClientFactoryOption<MeterRegistry> METER_REGISTRY =
            ClientFactoryOption.define("METER_REGISTRY", Metrics.globalRegistry);
```

It would be nice if this uses a `MeterRegistry` bean like that of the server version

**Comment by trustin** [comment]
Hey @kev1n11! Thanks for your feedback. The problem is we never had any Spring Boot integration for Armeria clients. We could probably expose a `ClientFactory` bean that's configured properly, e.g. `MeterRegistry`, so that a user can get it injected:

```java
@Configuration
public class MyConfiguration {
    private final MeterRegistry myMeterRegistry = ...;

    @Bean
    public MeterRegistry myMeterRegistry() {
        return myMeterRegistry;
    }

    @Bean
    public WebClient myWebClient(ClientFactory factory) {
        assert factory.meterRegistry() == myMeterRegistry;
        return WebClient.builder().factory(factory).build();
    }

    // We could also add a configurator interface similar to `ArmeriaServerConfigurator`.
    @Bean
    public ArmeriaClientConfigurator myClientConfigurator() {
        return clientFactoryBuilder -> {
            // More customization
            clientFactoryBuilder.tlsNoVerify();
        };
    }
}
```

However, it's probably not easy to provide the same level of customizability for `ClientFactory.ofDefault()`. What do you think?

**Comment by heowc** [comment]
I think creating an `ArmeriaClientConfigurator` is a good idea. 👍

In addition, what do you think of creating an auto-configuration class for Armeria client and making `WebClientBuilder` with bean?

I referenced [`spring-boot/RestaTemplateAutoConfiguration`](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/client/RestTemplateAutoConfiguration.java).

**Comment by trustin** [comment]
> In addition, what do you think of creating an auto-configuration class for Armeria client and making `WebClientBuilder` with bean?

You mean exposing `WebClient` without a base URL as a bean? That sounds fine to me.

**Comment by heowc** [comment]
We may need more settings, but for now I think this is enough.

```java
@Configuration
@ConditionalOnClass(WebClient.class)
@ConditionalOnMissingBean(WebClientBuilder.class)
public class ArmeriaClientAutoConfiguration {

    @Bean
    public WebClientBuilder webClientBuilder(Optional<List<ArmeriaClientConfigurator>> configurators) {
        final WebClientBuilder builder = WebClient.builder();

        configurators.ifPresent(
                it -> it.forEach(configurator -> configurator.configure(builder)));

        return builder;
    }

    @Bean
    @ConditionalOnBean(MeterRegistry.class)
    public ClientFactory clientFactory(MeterRegistry registry) {
        return ClientFactory.builder().meterRegistry(registry).build();
    }

    @Bean
    @ConditionalOnBean(ClientFactory.class)
    public ArmeriaClientConfigurator clientConfigurator(ClientFactory factory) {
        return builder -> builder.factory(factory);
    }
}
```

**Comment by trustin** [comment]
Sounds good to me, @heowc. Are you working on this already? 😄

**Comment by heowc** [comment]
No, I'm currently working on [a different issue](https://github.com/line/armeria/issues/3137).

If this feature doesn't need to be added right now, I'll add it after the spring boot dependency is updated to 2.4. We may have to do some unnecessary work to add that feature right now.
- See https://github.com/line/armeria/issues/3147

Also, I'm not sure if something else other than `MeterRegistry` should be set automatically. 🤔

**Comment by trustin** [comment]
> No, I'm currently working on [a different issue](https://github.com/line/armeria/issues/3137).

That's fine. You have every right to prioritize your contribution. 😉 

> Also, I'm not sure if something else other than `MeterRegistry` should be set automatically. 🤔

Would it make sense to add an interface like `ClientFactoryConfigurator` and use it for configuring the `ClientFactory` generated by our Spring Boot integration? (of course unless a user did not create his/her own `ClientFactory`.)

**Comment by heowc** [comment]
If what I understood is correct, Spring Boot made `ClientFactory` integrates only simple things (eg, metricRestry), and if this requires more functions, it is better to guide users to create and use `ClientFactory` themselves. What did you think?

**Comment by heowc** [comment]
I can afford to see this now. 😎

Can I work on it?

**Comment by ikhoon** [comment]
Sure, it is all yours. :-)
